File: translation_tools/translator.py
import json
import urllib.request
import urllib.parse
import time
import sys
import concurrent.futures

# Set stdout to fallback on error so console doesn't crash on emojis
sys.stdout.reconfigure(errors='replace')

missing_file = 'missing_translations.json'
lang_file = 'resources/lang/en.json'

with open(missing_file, 'r', encoding='utf-8') as f:
    missing = json.load(f)

with open(lang_file, 'r', encoding='utf-8') as f:
    existing = json.load(f)

def translate(text):
    if text in existing: return None
    if not text.strip(): return None
    url = "https://translate.googleapis.com/translate_a/single?client=gtx&sl=ar&tl=en&dt=t&q=" + urllib.parse.quote(text)
    req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
    try:
        with urllib.request.urlopen(req, timeout=5) as response:
            res = json.loads(response.read().decode('utf-8'))
            t_text = "".join([s[0] for s in res[0]])
            return text, t_text
    except Exception as e:
        return text, text

to_translate = [k for k in missing if k not in existing and k.strip()]
print(f"Remaining to translate: {len(to_translate)}")

count = 0
with concurrent.futures.ThreadPoolExecutor(max_workers=5) as executor:
    futures = {executor.submit(translate, text): text for text in to_translate}
    for future in concurrent.futures.as_completed(futures):
        res = future.result()
        if res:
            ar, en = res
            existing[ar] = en
            count += 1
            print(f"[{count}/{len(to_translate)}] translated.")
        
        if count % 20 == 0:
            with open(lang_file, 'w', encoding='utf-8') as f:
                json.dump(existing, f, ensure_ascii=False, indent=4)

with open(lang_file, 'w', encoding='utf-8') as f:
    json.dump(existing, f, ensure_ascii=False, indent=4)

print("Translation fully complete! Saved to en.json")

File: translation_tools/translator.php
<?php
$missing_file = 'missing_translations.json';
$lang_file = 'resources/lang/en.json';

$missing = json_decode(file_get_contents($missing_file), true);
$existing = json_decode(file_get_contents($lang_file), true) ?: [];

function translateBatch($texts) {
    if (empty($texts)) return [];
    
    // Using MyMemory translation API as it supports bulk or is faster for bulk sentences
    // Let's stick to Google Translate but fast
    $translatedList = [];
    foreach($texts as $text) {
        $url = "https://translate.googleapis.com/translate_a/single?client=gtx&sl=ar&tl=en&dt=t&q=" . urlencode($text);
        
        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, $url);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
        curl_setopt($ch, CURLOPT_USERAGENT, 'Mozilla/5.0');
        $response = curl_exec($ch);
        curl_close($ch);
        
        $res = json_decode($response, true);
        $translated_text = "";
        if (isset($res[0]) && is_array($res[0])) {
            foreach ($res[0] as $r) {
                $translated_text .= $r[0];
            }
        }
        $translatedList[$text] = empty($translated_text) ? $text : $translated_text;
        usleep(100000); // 0.1s
    }
    return $translatedList;
}

// Filter missing that are ALREADY in existing
$toTranslate = [];
foreach ($missing as $key) {
    if (!isset($existing[$key]) && trim($key) !== '') {
        $toTranslate[] = $key;
    }
}

echo "Remaining to translate: " . count($toTranslate) . " items...\n";

$count = 0;
foreach (array_chunk($toTranslate, 30) as $chunk) {
    $results = translateBatch($chunk);
    foreach ($results as $original => $english) {
        $existing[$original] = $english;
        $count++;
        echo "[$count/" . count($toTranslate) . "] $original => $english\n";
    }
    file_put_contents($lang_file, json_encode($existing, JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT));
}

echo "\nTranslation fully complete! Saved to en.json\n";
